src/coreclr/jit/hwintrinsiccodegenarm64.cpp

```diff
@@ -364,7 +364,44 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
         {
             assert(!hasImmediateOperand);
 
-            switch (intrin.numOperands)
+            size_t numOperands = intrin.numOperands;
+
+            // This handles optimizations for instructions that have
+            // an implicit 'zero' vector of what would be the second operand.
+            if ((numOperands == 2) && ((intrin.op2->IsVectorZero() && intrin.op2->isContained()) ||
+                                       (intrin.op1->IsVectorZero() && intrin.op1->isContained() &&
+                                        HWIntrinsicInfo::IsCommutative(intrin.id))))
+            {
+                assert(HWIntrinsicInfo::SupportsContainment(intrin.id));
+
+                if (intrin.op1->IsVectorZero() && intrin.op1->isContained() &&
+                    HWIntrinsicInfo::IsCommutative(intrin.id))
+                {
+                    // The intrinsic is commutative, swap the registers.
+                    assert(op1Reg == REG_NA);
+                    op1Reg = op2Reg;
+                    op2Reg = REG_NA;
+                }
```

Comment:
For x86/x64 we actually do the operand swap in lowering to help simplify codegen a bit.

For example:
https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/lowerxarch.cpp#L6113-L6119

Which then means in codegen we only ever have to consider `intrin.op2->isContained()`

It might be good to do that here too, but I'll defer to @echesakovMSFT

Reply:
I don't mind where we do the swap, but the one positive thing about doing it in codegen is that it isn't dependent on any specific intrinsic.

Reply:
I think moving the logic to be table-driven complicates things.
I would rather leave it as a special handling.

 > For x86/x64 we actually do the operand swap in lowering to help simplify codegen a bit.

I would do this too - this would slightly simplify the codegen part.

Reply:
Moving the swap to lowering is fine; it will simplify the logic in codegen though make it more dependent on specific intrinsics, but it only benefits `CompareEqual*` intrinsics so it makes sense to do it lowering.

> I think moving the logic to be table-driven complicates things.

Can you explain why this complicates things? It makes it easier to focus on the specific instruction rather than the intrinsic.

Reply:
> Can you explain why this complicates things? It makes it easier to focus on the specific instruction rather than the intrinsic.

You are adding special handling under `else` branch that before the change depended on number of operands of the intrinsic - but now it depends on `ins` (i.e. on a particular instruction opcode). This is not what I would think of as table-driven implementation. This should be abstracted from instructions and use more general categories (e.g. `SIMDByIndexedElement`, `ShiftLeftByImmediate` etc) for determining what path to take.

If you want to keep the logic table-driven, you might consider using combination of helper `SupportsContainment(ins)` and intrin.op2->isContained()` instead of the `switch` or, alternatively, introduce another category (e.g. you can see how this was done for `SIMDByIndexedElement` class).

Reply:
@tannergooding 

> Rather than needing to specially handle the same 'n' comparison in lowering and codegen, and ensuring they stay in sync/etc; we just annotate with a flag that it supports containment of zero and it implicitly lights up.

This is not what I was objecting to. Having a flag is fine. Moving the intrinsics to a dedicated category `SIMDCompare` would also be fine. Checking for a specific instruction opcode in a code (that was supposed to be general) is not.

Reply:
@echesakovMSFT, you're suggesting that something like:
```cpp
switch (numOperands)
{
   // ... existing code

    case 2:
    {
        if (intrin.SupportsContainmentZero() && intrin.op2->isContained())
        {
            assert(intrin.op2->isVectorZero());
        }
        else if (isRMW)
        // ... existingCode
    }

   // ... existing code
}
```

would be better here, right?

Reply:
(not necessarily exactly that, but something similar or along those lines)

Reply:
> but now it depends on ins (i.e. on a particular instruction opcode). This is not what I would think of as table-driven implementation.

That's fair and even in the xarch path, there is no branching on specific instructions as well.
I added a flag called "SupportsContainmentZero" and appropriately used the flag on the instrinsics we care about.

Reply:
> @echesakovMSFT, you're suggesting that something like:

@tannergooding Yes. 

Actually, I wouldn't focus that much on that the contained operand is 'zero`. So something like
```cpp
switch (numOperands)
{
   // ... existing code

    case 2:
    {
        // Lowering ensures that only op2 need to be checked for containment and will swap operands if needed.
        if (intrin.SupportsContainment() && intrin.op2->isContained())
        {

        }
        else if (isRMW)
        // ... existingCode
    }

   // ... existing code
}
```
would also work.

> That's fair and even in the xarch path, there is no branching on specific instructions as well.
I added a flag called "SupportsContainmentZero" and appropriately used the flag on the instrinsics we care about; so no more acting on specific instructions.

@TIHan Thanks!